### ISSUE TITLE
[Identity][Antiforgery] Invalid antiforgery token error

### ISSUE THREAD

**Issue by javiercn** [aspnet/AspNetCore#11485]
_From @markphillips100_

>@javiercn I consistently get this issue due to a particular use case that includes >using AspNetCore Identity as the user accounts provider for a process hosting >IdentityServer4.
>
>I can reproduce the error with a standard .net core 3.0-preview5 web application >using local user accounts and the ForgotPassword/Reset Password process.
>
>The template's process works by directing the user back to log in page from the >Reset Password confirmation page, whereas in my case I expect the user to click a >link back to log in page from the original Forgot Password confirmation page. This >is due to the fact that only the original oauth flow can redirect back to the app >requesting authentication.
>
>I capitulate that AspNetCore Identity's use case leads the user to avoid the error >and mine leads the user into the error potentially occurring. However, I see no >alternative to the use-case that I require.
>
>Is there anyway to trap this specific exception and re-execute a GET request >perhaps?
>
>Steps to reproduce
>You can reproduce both working (standard use-case) and failing (my use-case) use >cases consistently using a basic dotnet web app template with local user accounts >and email reset password use case.
>
>I followed the docs here to create the basic web app with user accounts, and then >setup sendgrid for the email. I already had registered an account prior to setting >the sendgrid email so I set the IsEmailConfirmed property in the database to True.
>
>Environment setup
>DotNet: .net core 3.0-preview5
>Server: Windows 10 and Kestrel web server
>DevTools: VS2019 v16.1.3
>
>Working ForgotPassword/ResetPassword process
>
>Launch app
>Click Login.
>Click Forgot your password?
>Enter valid email and submit.
>Forgot Password Confirmation page shown. (leave this page open).
>Receive email and click link in email. Fill out Reset Password form correctly and >submit.
>Reset Password Confirmation page shown with a link to login.
>At this point, you can either click the Go to Login link on Reset Password >Confirmation page or click the Login link in the menu on the Forgot Password >Confirmation page.
>Login page displayed and user can log in with no problems.
>Failing ForgotPassword/ResetPassword process
>
>Launch app
>Click Login.
>Click Forgot your password?
>Enter valid email and submit.
>Forgot Password Confirmation page shown.
>Click the Login link in menu
>Receive email and click link in email. Fill out Reset Password form correctly and >submit.
>Reset Password Confirmation page shown with a link to login. Do not click this link
>Fill out Login page and submit.
>AntiForgery error occurs.

**Comment by markphillips100** [comment]
@javiercn thanks for opening a new issue for me.

**Comment by javiercn** [comment]
I believe your case is expected and by design with regards to antiforgery for the following reasons. What happens is as follows:
* Browser A goes to the login page, up to step 6 in `Click the Login link in menu` with an antiforgery token for an unauthenticated user
* Browser B gets opened when the user clicks the link. User introduces the new password and hits submit.
  * At this point the user gets authenticated, which means its identity changes and the old antiforgery token becomes invalid due to the user having a new Identity.
* User goes back to browser A enters login and password and hits submit and then you get the antiforgery error (as the identity of the user changed).

Am I interpreting your steps correctly? If you have multiple browsers at play tagging them on the steps will help understanding the flow better.

I'm not sure that there is anything you can do in this case that would allow you to avoid this error. Antiforgery is working as expected in this case, it needs to match the identity of the currently authenticated user. (Which after submitting the reset password form changes.

What happens in your original login form is that the antiforgery token rendered in that page becomes obsolete the moment the user resets its password on the second browser instance.

@HaoK After reset password succeeds the user is authenticated, isn't it?
@blowdart This if anything would be an issue with the flows in identity, not on antiforgery, so I leave it up to you what you want to do here, although my take is that nothing can probably be done.

**Comment by markphillips100** [comment]
@javiercn You are interpreting my steps correctly.  

Specifically, my **Browser A** is hosted by a mobile app (using SafariViewController/GoogleChromeTabs).  The mobile app user is directed to the login from within the app.  The reset password email link opens in a **separate system mobile browser** so is **Browser B** in your description.  Note: both browsers are using the system browser and share cookies.

After **Browser B** submits the reset password form, the mobile user likely has the login page still open in **Browser A** (as that's where they are expected to login) and so would naturally attempt a login once finished resetting the password.  The user cannot login via a link from the reset password confirmation page (in my use case).

Like you I don't believe this is necessarily anything to do with antiforgery, i.e. that is working as expected.    That's why I'm just wondering if there's some (safe) way of capturing the exception and reloading the page.

**Comment by markphillips100** [comment]
I guess one way of resolving it would be to host the reset password page in a completely separate host (different middleware, different antiforgery cookie) but seems like a hammer approach.

**Comment by javiercn** [comment]
> Note: both browsers are using the system browser and share cookies.

That's exactly the issue. Antiforgery uses a request token + a cookie, when you log-in on browser B after reset password succeeds, the antiforgery cookie gets updated with the new user identity and causes the request token in browser A to become invalid.

You might be able to workaround this by refreshing the login page preemptively somehow, but I can't think of a good way to do so.

You might also override the login page and extend it to (very carefully) disable automatic antiforgery validation, perform it yourself and rerender the login page when an antiforgery error happens.

Under no circumstance should you ever retry the request without new user input or simply ignore the error, as that would open you to CSRF vulnerabilities.

Do you have any data on how often do your users run into this problem? It might help us in the future to change something here, but as it stands today is by design.

**Comment by markphillips100** [comment]
I don't think I'll attempt to disable the antiforgery and workaround it but thanks for the tip about not retrying the request.

I have no user data at the moment as this is still in development.  However, just my own use of the mobile app use case proved that this surfaced fairly simply.

In the meantime, I've just created a copy of the ResetPassword page in a separate webhost project - my aspnetcore identity Account pages are hosted separately from the Manage pages.  This setup means the login page's antiforgery token isn't invalidated.

I can live with this setup for now but I might experiment with overriding the anitforgery cookie name for specific request paths and thus allow me to keep the ResetPassword page where it belongs. Unless you can see an obvious problem with this approach?

**Comment by markphillips100** [comment]
Not that I know if overriding the cookie name would actually work or not but I failed anyway.  I tried to override the cookie name per path (using IOptionsMonitor) but it seems the internal DefaultAntiforgery type is registered as a singleton and it's constructor accesses the injected IOptions<AntiforgeryOptions> parameter's Value property.  So basically, my options are called once only.

I shall have to live with hosting ResetPassword page separately.

Thanks for your help @javiercn and at least broadening my understanding of Antiforgery.
